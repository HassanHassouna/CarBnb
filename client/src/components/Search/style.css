.container__searchBar {
  display: flex;
  height: 40vh;
  justify-content: center;
  background-size: cover;
  background-repeat: no-repeat;
  background-position: 0 40%;
  background-image: url("https://live.staticflickr.com/7097/7230382860_ec128631cc_b.jpg");
}

.vertical_line {
  width: 1px;
  height: 80%;
  background-color: #e0e0e0;
}
.searchContainer {
  display: flex;
  margin-top: 3vh;
  width: 60%;
  height: 6vh;
  border-radius: 35px;
  align-items: center;
  background-color: white;
  box-shadow: 0px 0px 8px 0px rgba(0, 0, 0, 0.15);
}
