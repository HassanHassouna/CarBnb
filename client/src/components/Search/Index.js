import React from "react"
import { useDispatch } from "react-redux"
import Fab from "@mui/material/Fab"
import Box from "@mui/material/Box"
import TextField from "@mui/material/TextField"
import Autocomplete from "@mui/material/Autocomplete"
import LocationOnIcon from "@mui/icons-material/LocationOn"
import Grid from "@mui/material/Grid"
import Typography from "@mui/material/Typography"
import parse from "autosuggest-highlight/parse"
import throttle from "lodash/throttle"
import Datee from "./Date/Index"
import Time from "./Time/Index"
import SearchIcon from "@mui/icons-material/Search"
import IconButton from "@mui/material/IconButton"
import "./style.css"
import { search } from "../../app/actions/search-actions"

const GOOGLE_MAPS_API_KEY = 'AIzaSyAsJrza-9qgAdE5FUD2f26prJwV9vCt7wA';

function loadScript(src, id) {
  return new Promise((resolve, reject) => {
    const script = document.createElement('script');
    script.setAttribute('async', '');
    script.setAttribute('id', id);
    script.src = src;
    script.onload = resolve;
    script.onerror = reject;
    document.head.appendChild(script);
  });
}

const autocompleteService = {
  current: null,
};

export default function Search() {
  const dispatch = useDispatch();
  const [value, setValue] = React.useState(null);
  const [inputValue, setInputValue] = React.useState('');
  const [options, setOptions] = React.useState([]);
  const loaded = React.useRef(false);
  const whereRef = React.useRef();
  const fromRef = React.useRef();
  const untilRef = React.useRef();
  const timeToPickRef = React.useRef();
  const timeToDropRef = React.useRef();

  const searchHandler = () => {
    console.log("fromRef", fromRef.current.value, typeof fromRef.current.value)
    console.log(
      "timeToPick",
      timeToPickRef.current.value,
      typeof timeToPickRef.current.value
    )
    const yearFrom = fromRef.current.value.split("/")[2]
    const monthFrom = fromRef.current.value.split("/")[0]
    const dayFrom = fromRef.current.value.split("/")[1]
    const hoursFrom = timeToPickRef.current.value.split(":")[0]
    const minutesFrom = timeToPickRef.current.value.split(":")[1]

    const yearUntil = untilRef.current.value.split("/")[2]
    const monthUntil = untilRef.current.value.split("/")[0]
    const dayUntil = untilRef.current.value.split("/")[1]
    const hoursUntil = timeToDropRef.current.value.split(":")[0]
    const minutesUntil = timeToDropRef.current.value.split(":")[1]

    const searchDataObject = {
      location: whereRef.current.value,
      start_order: new Date(
        yearFrom,
        monthFrom,
        dayFrom,
        hoursFrom,
        minutesFrom
      ),
      end_order: new Date(
        yearUntil,
        monthUntil,
        dayUntil,
        hoursUntil,
        minutesUntil
      ),
    }
    dispatch(search(searchDataObject))
  }

  if (typeof window !== 'undefined' && !loaded.current) {
    if (!document.querySelector('#google-maps')) {
      loadScript(
        `https://maps.googleapis.com/maps/api/js?key=${GOOGLE_MAPS_API_KEY}&libraries=places`,
        document.body,
        'google-maps'
      ).then(() => {
        autocompleteService.current =
          new window.google.maps.places.AutocompleteService();

        loaded.current = true;
      });
    }
  }

  const fetch = React.useMemo(
    () =>
      throttle((request, callback) => {
        autocompleteService.current.getPlacePredictions(request, callback);
      }, 200),
    []
  );

  React.useEffect(() => {
    let active = true;
    if (!autocompleteService.current) {
      return undefined;
    }

    if (inputValue === '') {
      setOptions(value ? [value] : []);
      return undefined;
    }

    fetch({ input: inputValue }, (results) => {
      if (active) {
        let newOptions = [];

        if (value) {
          newOptions = [value];
        }

        if (results) {
          newOptions = [...newOptions, ...results];
        }

        setOptions(newOptions);
      }
    });

    return () => {
      active = false;
    };
  }, [value, inputValue, fetch]);

  return (
    <div className="container__searchBar">
      <div className="searchContainer">
        <Autocomplete
          id="google-map-demo"
          sx={{ width: 400 }}
          getOptionLabel={(option) =>
            typeof option === 'string' ? option : option.description
          }
          filterOptions={(x) => x}
          options={options}
          autoComplete
          includeInputInList
          filterSelectedOptions
          value={value}
          onChange={(event, newValue) => {
            setOptions(newValue ? [newValue, ...options] : options);
            setValue(newValue);
          }}
          onInputChange={(event, newInputValue) => {
            setInputValue(newInputValue);
          }}
          renderInput={(params) => (
            <TextField
              inputRef={whereRef}
              {...params}
              fullWidth
              placeholder="Where?"
            />
          )}
          renderOption={(props, option) => {
            const matches =
              option.structured_formatting.main_text_matched_substrings;
            const parts = parse(
              option.structured_formatting.main_text,
              matches.map((match) => [
                match.offset,
                match.offset + match.length,
              ])
            );

            return (
              <li {...props}>
                <Grid container alignItems="center">
                  <Grid item>
                    <Box
                      component={LocationOnIcon}
                      sx={{ color: 'text.secondary', mr: 2 }}
                    />
                  </Grid>
                  <Grid item xs>
                    {parts.map((part, index) => (
                      <span
                        key={index}
                        style={{
                          fontWeight: part.highlight ? 700 : 400,
                        }}
                      >
                        {part.text}
                      </span>
                    ))}

                    <Typography variant="body2" color="text.secondary">
                      {option.structured_formatting.secondary_text}
                    </Typography>
                  </Grid>
                </Grid>
              </li>
            );
          }}
        />
        <div className="vertical_line"></div>
        <Datee ref={fromRef} label="From" />
        <Time ref={timeToPickRef} />
        <div className="vertical_line"></div>
        <Datee ref={untilRef} label="Until" />
        <Time ref={timeToDropRef} />
        <IconButton
          color="primary"
          aria-label="upload picture"
          component="label"
          onClick={searchHandler}
        >
          <Box>
            <Fab size="small" color="primary" aria-label="search">
              <SearchIcon />
            </Fab>
          </Box>
        </IconButton>
      </div>
    </div>
  );
}
