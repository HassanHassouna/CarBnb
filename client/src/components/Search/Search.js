import { useState, useRef, useMemo, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useNavigate } from "react-router-dom";
import Fab from "@mui/material/Fab";
import Box from "@mui/material/Box";
import TextField from "@mui/material/TextField";
import Autocomplete from "@mui/material/Autocomplete";
import LocationOnIcon from "@mui/icons-material/LocationOn";
import Grid from "@mui/material/Grid";
import Typography from "@mui/material/Typography";
import parse from "autosuggest-highlight/parse";
import throttle from "lodash/throttle";
import Datee from "./Date/Index";
import Time from "./Time/Index";
import SearchIcon from "@mui/icons-material/Search";
import IconButton from "@mui/material/IconButton";
import "./style.css";
import { search } from "../../app/actions/search-actions";
import LoadingSpinner from "../loadingSpinner/LoadingSpinner";
const GOOGLE_MAPS_API_KEY = "AIzaSyAsJrza-9qgAdE5FUD2f26prJwV9vCt7wA";

function loadScript(src, id) {
  return new Promise((resolve, reject) => {
    const script = document.createElement("script");
    script.setAttribute("async", "");
    script.setAttribute("id", id);
    script.src = src;
    script.onload = resolve;
    script.onerror = reject;
    document.head.appendChild(script);
  });
}

const autocompleteService = {
  current: null,
};

export default function Search() {
  const loading = useSelector((state) => state.viewSlice.isLoading);
  const isError = useSelector((state) => state.viewSlice.isError);
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const [value, setValue] = useState(null);
  const [inputValue, setInputValue] = useState("");
  const [options, setOptions] = useState([]);
  const loaded = useRef(false);
  const whereRef = useRef();
  const fromRef = useRef();
  const untilRef = useRef();
  const timeToPickRef = useRef();
  const timeToDropRef = useRef();

  const searchHandler = () => {
    const dateFrom = fromRef.current.value.split("/");
    const dateUntil = untilRef.current.value.split("/");
    const timeFrom = timeToPickRef.current.value.split(":");
    const timeUntil = timeToDropRef.current.value.split(":");

    const yearFrom = dateFrom[2];
    const monthFrom = dateFrom[0];
    const dayFrom = dateFrom[1];
    const hoursFrom = timeFrom[0];
    const minutesFrom = timeFrom[1];

    const yearUntil = dateUntil[2];
    const monthUntil = dateUntil[0];
    const dayUntil = dateUntil[1];
    const hoursUntil = timeUntil[0];
    const minutesUntil = timeUntil[1];

    const searchDataObject = {
      location: whereRef.current.value,
      start_order: new Date(
        yearFrom,
        monthFrom,
        dayFrom,
        hoursFrom,
        minutesFrom
      ),
      end_order: new Date(
        yearUntil,
        monthUntil,
        dayUntil,
        hoursUntil,
        minutesUntil
      ),
    };
    // validate data
    if (searchDataObject.location === "") {
      whereRef.current.focus();
      whereRef.current.placeholder = "Please enter a location";
      return;
    } else if (timeToPickRef.current.value === "") {
      timeToPickRef.current.focus();
      alert("Please enter a time");
    } else if (timeToDropRef.current.value === "") {
      timeToDropRef.current.focus();
      alert("Please enter a time");
    } else {
      dispatch(search(searchDataObject));
      navigate("/searchResult");
    }
  };

  if (typeof window !== "undefined" && !loaded.current) {
    if (!document.querySelector("#google-maps")) {
      loadScript(
        `https://maps.googleapis.com/maps/api/js?key=${GOOGLE_MAPS_API_KEY}&libraries=places`,
        document.body,
        "google-maps"
      ).then(() => {
        autocompleteService.current =
          new window.google.maps.places.AutocompleteService();

        loaded.current = true;
      });
    }
  }

  const fetch = useMemo(
    () =>
      throttle((request, callback) => {
        autocompleteService.current.getPlacePredictions(request, callback);
      }, 200),
    []
  );

  useEffect(() => {
    let active = true;

    if (!autocompleteService.current) {
      return undefined;
    }

    if (inputValue === "") {
      setOptions(value ? [value] : []);
      return undefined;
    }

    fetch({ input: inputValue }, (results) => {
      if (active) {
        let newOptions = [];

        if (value) {
          newOptions = [value];
        }

        if (results) {
          newOptions = [...newOptions, ...results];
        }

        setOptions(newOptions);
      }
    });

    return () => {
      active = false;
    };
  }, [value, inputValue, fetch]);

  if (loading) {
    return <LoadingSpinner />;
  }
  
  if (isError) {
    return <div>Error</div>;
  }

  return (
    <div className="container__searchBar">
      <div className="searchContainer">
        <Autocomplete
          id="google-map-demo"
          sx={{ width: 400 }}
          getOptionLabel={(option) =>
            typeof option === "string" ? option : option.description
          }
          filterOptions={(x) => x}
          options={options}
          autoComplete
          includeInputInList
          filterSelectedOptions
          value={value}
          onChange={(event, newValue) => {
            setOptions(newValue ? [newValue, ...options] : options);
            setValue(newValue);
          }}
          onInputChange={(event, newInputValue) => {
            setInputValue(newInputValue);
          }}
          renderInput={(params) => (
            <TextField
              variant="outlined"
              inputRef={whereRef}
              {...params}
              fullWidth
              placeholder="Where?"
              sx={{
                "& .css-1d3z3hw-MuiOutlinedInput-notchedOutline": {
                  borderStyle: "none",
                },
              }}
            />
          )}
          renderOption={(props, option) => {
            const matches =
              option.structured_formatting.main_text_matched_substrings;
            const parts = parse(
              option.structured_formatting.main_text,
              matches.map((match) => [
                match.offset,
                match.offset + match.length,
              ])
            );

            return (
              <li {...props}>
                <Grid container alignItems="center">
                  <Grid item>
                    <Box
                      component={LocationOnIcon}
                      sx={{
                        color: "text.secondary",
                        mr: 2,
                      }}
                    />
                  </Grid>
                  <Grid item xs>
                    {parts.map((part, index) => (
                      <span
                        key={index}
                        style={{
                          fontWeight: part.highlight ? 700 : 400,
                        }}
                      >
                        {part.text}
                      </span>
                    ))}

                    <Typography variant="body2" color="text.secondary">
                      {option.structured_formatting.secondary_text}
                    </Typography>
                  </Grid>
                </Grid>
              </li>
            );
          }}
        />
        <div className="vertical_line"></div>
        <Datee ref={fromRef} label="From" />
        <Time ref={timeToPickRef} />
        <div className="vertical_line"></div>
        <Datee ref={untilRef} label="Until" />
        <Time ref={timeToDropRef} />
        <IconButton
          color="primary"
          aria-label="upload picture"
          component="label"
          onClick={searchHandler}
        >
          <Box>
            <Fab size="small" color="primary" aria-label="search">
              <SearchIcon />
            </Fab>
          </Box>
        </IconButton>
      </div>
    </div>
  );
}
