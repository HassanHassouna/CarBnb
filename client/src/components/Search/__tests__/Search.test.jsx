import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import Search from '../Search';
import { Provider } from 'react-redux';
import { store } from '../../../app/store';
import { BrowserRouter as Router } from 'react-router-dom';

describe('Search', () => {
  test('should render with no parameters', () => {
    const { search } = render(
      <Provider store={store}>
        <Router>
          <Search />
        </Router>
      </Provider>
    );

    expect(search).toMatchSnapshot();
  });
});
