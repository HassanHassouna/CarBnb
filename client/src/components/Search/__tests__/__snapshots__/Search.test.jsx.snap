// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render with no parameters 1`] = `undefined`;
